src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -67,6 +64,211 @@ case class Task(
+    * and is currently not launched.
+    *
+    * @param taskId The task Id
+    * @param reservation Information about the reserved resources and persistent volumes
+    */
+  case class Reserved(
+      taskId: Task.Id,
+      agentInfo: AgentInfo,
+      state: Reserved.State,
+      reservation: ReservationWithVolumes) extends Task {
+
+    private[this] val log = LoggerFactory.getLogger(getClass)
+
+    override val reservationWithVolumes: Option[ReservationWithVolumes] = Some(reservation)
+
+    override val launched: Option[Launched] = None
+
+    override def update(update: TaskStateOp): TaskStateChange = update match {
+      case TaskStateOp.Launch(appVersion, status, networking) =>
+        TaskStateChange.Update(LaunchedOnReservation(
+          taskId, agentInfo, appVersion, status, networking, reservation))
+
+      case TaskStateOp.Timeout =>
+        TaskStateChange.Expunge
+
+      // failure case: MesosUpdate
+      case _: TaskStateOp.MesosUpdate =>
+        val msg = s"Unable to handle MesosUpdate op on Reserved task $taskId"
+        log.error(msg)
+        TaskStateChange.Failure(msg)
+    }
+  }
+
+  object Reserved {
+
+    sealed trait State {
+      /** Defines when this state should time out and for which reason */
+      def timeout: Option[Timeout]
+    }
+    object State {
+      /** A newly reserved Task */
+      case class New(timeout: Option[Timeout]) extends State
+      /** A ResidentTask that has been running before but terminated and can be relaunched */
+      case class Suspended(timeout: Option[Timeout]) extends State
+      /** A ResidentTask whose reservation and persistent volumes are being destroyed */
+      case class Garbage(timeout: Option[Timeout]) extends State
+      /** An unknown task created because of unknown reservations/persistent volumes */
+      case class Unknown(timeout: Option[Timeout]) extends State
+    }
+
+    /**
+      * A timeout that eventually leads to a state transition
+      * @param initiated When this timeout was setup
+      * @param deadline When this timeout should become effective
+      * @param reason The reason why this timeout was set up
+      */
+    case class Timeout(initiated: Timestamp, deadline: Timestamp, reason: Timeout.Reason)
+
+    object Timeout {
+      sealed trait Reason
+      object Reason {
+        /** A timeout because the task could not be relaunched */
+        case object RelaunchEscalationTimeout extends Reason
+        /** A timeout because we got no ack for reserved resources or persistent volumes */
+        case object ReservationTimeout extends Reason
+      }
+    }
+  }
+
+  case class LaunchedOnReservation(
+      taskId: Task.Id,
+      agentInfo: AgentInfo,
+      appVersion: Timestamp,
+      status: Status,
+      networking: Networking,
+      reservation: ReservationWithVolumes) extends Task {
+
+    private[this] val log = LoggerFactory.getLogger(getClass)
```

Comment:
Put the logger into the companion object. No need to resolve the logger for every task instance.